### ISSUE TITLE
Prisma generate randomly fails on Ubuntu due to missing internal .so `libquery_engine-debian-openssl-1.1.x.so.node`

### ISSUE THREAD

**Issue by jakeleventhal** [prisma/prisma#19124]
### Bug description

Here is my package.json in the root of my repo:
```json
{
  "devDependencies": {
    "@prisma/client": "4.13.0",
    "pnpm": "8.4.0",
    "prettier": "2.8.8",
    "prisma": "4.13.0",
    "prisma-json-types-generator": "2.3.1",
    "ts-node": "10.9.1",
    "typescript": "5.0.4",
    "turbo": "1.9.3",
    "vercel": "29.0.0"
  },
  "scripts": {
    "postinstall": "pnpm run -r --parallel prisma-generate"
  }
}
```
I have two database packages with `prisma-generate` scripts that look like so:
```json
"prisma-generate": "prisma generate --schema=./prisma/schema.prisma"
```

In my GitHub actions, when I run `pnpm install`, I randomly (and only sometimes) get the following error:
```
. postinstall$ pnpm run -r --parallel prisma-generate
. postinstall: Scope: 26 of 27 workspace projects
. postinstall: packages/ec-database prisma-generate$ prisma generate --schema=./prisma/schema.prisma
. postinstall: packages/pl-database prisma-generate$ prisma generate --schema=./prisma/schema.prisma
. postinstall: packages/ec-database prisma-generate: Prisma schema loaded from prisma/schema.prisma
. postinstall: packages/pl-database prisma-generate: Prisma schema loaded from prisma/schema.prisma
. postinstall: packages/pl-database prisma-generate: Error: ENOENT: no such file or directory, stat '/home/runner/work/Ecominate/Ecominate/node_modules/.pnpm/prisma@4.13.0/node_modules/prisma/libquery_engine-debian-openssl-1.1.x.so.node'
. postinstall: packages/pl-database prisma-generate: Failed
. postinstall: undefined
. postinstall: /home/runner/work/Ecominate/Ecominate/packages/pl-database:
. postinstall:  ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @ecominate/pl-database@1.0.0 prisma-generate: `prisma generate --schema=./prisma/schema.prisma`
. postinstall: Exit status 1
. postinstall: Failed
 ELIFECYCLE  Command failed with exit code 1.
```

### How to reproduce

See above. Can't consistently reproduce

### Expected behavior

No errors

### Prisma information

See above

### Environment & setup

- OS: `runs-on: ubuntu-latest` from my GH action
- Database: PostgreSQL
- Node.js version: `18.12.1`

### Prisma Version

```
4.13.0
```

**Comment by jakeleventhal** [comment]
I have 7 build jobs that run in parallel. I would say 9/10 at least one of my jobs fails with this error in GitHub Actions.

**Comment by marc2332** [comment]
I just solved this issue by upgrading to `4.14.1`, maybe that will help you :))

**Comment by darthmaim** [comment]
I have a project with 2 schemas that get generated in multiple build jobs, and I have to restart about half of the CI builds because one of them fails with this error…

Here is a log

```text
> prisma:generate
> prisma generate --schema=prisma/legacy.prisma

Environment variables loaded from .env
Prisma schema loaded from prisma/legacy.prisma
node:events:491
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open '/home/runner/work/gw2treasures.com/gw2treasures.com/node_modules/prisma/libquery_engine-debian-openssl-1.1.x.so.node'
Emitted 'error' event on ReadStream instance at:
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/home/runner/work/gw2treasures.com/gw2treasures.com/node_modules/prisma/libquery_engine-debian-openssl-1.1.x.so.node'
}

Node.js v18.16.0
npm ERR! Lifecycle script `prisma:generate` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: legacy-importer 
npm ERR!   at location: /home/runner/work/gw2treasures.com/gw2treasures.com/apps/legacy-importer 
ERROR: command finished with error: command (/home/runner/work/gw2treasures.com/gw2treasures.com/apps/legacy-importer) npm run prisma:generate exited (1)
command (/home/runner/work/gw2treasures.com/gw2treasures.com/apps/legacy-importer) npm run prisma:generate exited (1)
```

And here is the run: https://github.com/GW2Treasures/gw2treasures.com/actions/runs/5405022539/jobs/9820002707?pr=335#step:6:30

If there is anything I can do to help to narrow down the cause please let me know.

**Comment by jakeleventhal** [comment]
Same thing here @darthmaim

**Comment by hiroingk** [comment]
I am getting the same error with prisma 4.16.2.

**Comment by Jolg42** [comment]
Looking at @darthmaim [comment](https://github.com/prisma/prisma/issues/19124#issuecomment-1612047273)
I think the `ReadStream` error comes from
https://github.com/prisma/prisma/blob/main/packages/fetch-engine/src/getHash.ts

The current workaround that you found is to do some retries https://github.com/GW2Treasures/gw2treasures.com/pull/389 which is not great for sure.

@darthmaim and others, could you set the en var `DEBUG=prisma*` and share the logs?
It would give us a better idea of what's happening, from a first look it seems like there is a race condition to me.

**Comment by darthmaim** [comment]
I've added `DEBUG=prisma*` to my CI script.

I restarted the job 10 times and didn't get a single error, usually I only get 3 runs or so in a row without the error.. So either I'm really lucky (or unlucky), or the additional debug logs somehow prevent a race condition. I've hit the error multiple earlier today before adding the debug logs.

I will report back when the error comes back with some additional logs

**Comment by Jolg42** [comment]
Thanks for the update and very curious to see if it comes back 👀

**Comment by jakeleventhal** [comment]
@Jolg42 its very flakey. As @darthmaim mentioned, I get this sometimes 5 times in a row. Sometimes 20 times without occurring.

**Comment by darthmaim** [comment]
Just got the error again, here are the debug logs:

```raw
> DEBUG=prisma* prisma generate --schema=prisma/legacy.prisma
  
  2023-07-29T11:[19](https://github.com/GW2Treasures/gw2treasures.com/actions/runs/5700125504/job/15449673348#step:7:20):59.391Z prisma:engines  binaries to download libquery-engine, schema-engine
  [20](https://github.com/GW2Treasures/gw2treasures.com/actions/runs/5700125504/job/15449673348#step:7:21)23-07-29T11:19:59.395Z prisma:get-platform  Found distro info:
  {
    "targetDistro": "debian",
    "familyDistro": "debian",
    "originalDistro": "ubuntu"
  }
  2023-07-29T11:19:59.395Z prisma:get-platform  Trying platform-specific paths for "debian" (and "ubuntu")
  2023-07-29T11:19:59.397Z prisma:get-platform  Found libssl.so file using platform-specific paths: libssl.so
  2023-07-29T11:19:59.397Z prisma:get-platform  The parsed libssl version is: undefined
  2023-07-29T11:19:59.397Z prisma:get-platform  Falling back to "ldconfig" and other generic paths
  2023-07-29T11:19:59.435Z prisma:get-platform  Command "ldconfig -p | sed "s/.*=>s*//" | sed "s|.*/||" | grep libssl | sort | grep -v "libssl.so.0"" successfully returned "libssl.so
  libssl.so.1.1
  libssl.so.3
  libssl3.so
  "
  2023-07-29T11:19:59.435Z prisma:get-platform  Found libssl.so file using "ldconfig" or other generic paths: libssl.so
  libssl.so.1.1
  libssl.so.3
  libssl3.so
  
  2023-07-29T11:19:59.435Z prisma:get-platform  The parsed libssl version is: 1.1.x
  2023-07-29T11:19:59.441Z prisma:get-platform  Found distro info:
  {
    "targetDistro": "debian",
    "familyDistro": "debian",
    "originalDistro": "ubuntu"
  }
  2023-07-29T11:19:59.441Z prisma:get-platform  Trying platform-specific paths for "debian" (and "ubuntu")
  2023-07-29T11:19:59.442Z prisma:get-platform  Found libssl.so file using platform-specific paths: libssl.so
  2023-07-29T11:19:59.442Z prisma:get-platform  The parsed libssl version is: undefined
  2023-07-29T11:19:59.442Z prisma:get-platform  Falling back to "ldconfig" and other generic paths
  2023-07-29T11:19:59.464Z prisma:get-platform  Command "ldconfig -p | sed "s/.*=>s*//" | sed "s|.*/||" | grep libssl | sort | grep -v "libssl.so.0"" successfully returned "libssl.so
  libssl.so.1.1
  libssl.so.3
  libssl3.so
  "
  2023-07-29T11:19:59.464Z prisma:get-platform  Found libssl.so file using "ldconfig" or other generic paths: libssl.so
  libssl.so.1.1
  libssl.so.3
  libssl3.so
  
  2023-07-29T11:19:59.464Z prisma:get-platform  The parsed libssl version is: 1.1.x
  2023-07-29T11:19:59.924Z prisma:loadEnv  project root found at /home/runner/work/gw2treasures.com/gw2treasures.com/apps/legacy-importer/package.json
  2023-07-29T11:19:59.931Z prisma:tryLoadEnv  Environment variables loaded from /home/runner/work/gw2treasures.com/gw2treasures.com/apps/legacy-importer/.env
  Environment variables loaded from .env
  Prisma schema loaded from prisma/legacy.prisma
  2023-07-29T11:19:59.953Z prisma:get-platform  Found distro info:
  {
    "targetDistro": "debian",
    "familyDistro": "debian",
    "originalDistro": "ubuntu"
  }
  2023-07-29T11:19:59.953Z prisma:get-platform  Trying platform-specific paths for "debian" (and "ubuntu")
  2023-07-29T11:19:59.955Z prisma:get-platform  Found libssl.so file using platform-specific paths: libssl.so
  2023-07-29T11:19:59.955Z prisma:get-platform  The parsed libssl version is: undefined
  2023-07-29T11:19:59.955Z prisma:get-platform  Falling back to "ldconfig" and other generic paths
  2023-07-29T11:19:59.984Z prisma:get-platform  Command "ldconfig -p | sed "s/.*=>s*//" | sed "s|.*/||" | grep libssl | sort | grep -v "libssl.so.0"" successfully returned "libssl.so
  libssl.so.1.1
  libssl.so.3
  libssl3.so
  "
  2023-07-29T11:19:59.984Z prisma:get-platform  Found libssl.so file using "ldconfig" or other generic paths: libssl.so
  libssl.so.1.1
  libssl.so.3
  libssl3.so
  
  2023-07-29T11:19:59.984Z prisma:get-platform  The parsed libssl version is: 1.1.x
  2023-07-29T11:19:59.993Z prisma:get-platform  Found distro info:
  {
    "targetDistro": "debian",
    "familyDistro": "debian",
    "originalDistro": "ubuntu"
  }
  2023-07-29T11:19:59.993Z prisma:get-platform  Trying platform-specific paths for "debian" (and "ubuntu")
  2023-07-29T11:19:59.994Z prisma:get-platform  Found libssl.so file using platform-specific paths: libssl.so
  2023-07-29T11:19:59.994Z prisma:get-platform  The parsed libssl version is: undefined
  2023-07-29T11:19:59.994Z prisma:get-platform  Falling back to "ldconfig" and other generic paths
  2023-07-29T11:20:00.017Z prisma:get-platform  Command "ldconfig -p | sed "s/.*=>s*//" | sed "s|.*/||" | grep libssl | sort | grep -v "libssl.so.0"" successfully returned "libssl.so
  libssl.so.1.1
  libssl.so.3
  libssl3.so
  "
  2023-07-29T11:20:00.017Z prisma:get-platform  Found libssl.so file using "ldconfig" or other generic paths: libssl.so
  libssl.so.1.1
  libssl.so.3
  libssl3.so
  
  2023-07-29T11:20:00.017Z prisma:get-platform  The parsed libssl version is: 1.1.x
  2023-07-29T11:20:00.110Z prisma:download  copying /home/runner/.cache/prisma/master/6b0aef69b7cdfc787f822ecd7cdc76d5f1991584/debian-openssl-1.1.x/libquery-engine to /home/runner/work/gw2treasures.com/gw2treasures.com/node_modules/prisma/libquery_engine-debian-openssl-1.1.x.so.node
  node:events:492
        throw er; // Unhandled 'error' event
        ^
  
  Error: ENOENT: no such file or directory, open '/home/runner/work/gw2treasures.com/gw2treasures.com/node_modules/prisma/libquery_engine-debian-openssl-1.1.x.so.node'
  Emitted 'error' event on ReadStream instance at:
      at emitErrorNT (node:internal/streams/destroy:151:8)
      at emitErrorCloseNT (node:internal/streams/destroy:116:3)
      at process.processTicksAndRejections (node:internal/process/task_queues:82:[21](https://github.com/GW2Treasures/gw2treasures.com/actions/runs/5700125504/job/15449673348#step:7:22)) {
    errno: -2,
    code: 'ENOENT',
    syscall: 'open',
    path: '/home/runner/work/gw2treasures.com/gw2treasures.com/node_modules/prisma/libquery_engine-debian-openssl-1.1.x.so.node'
  }
  
  Node.js v20.5.0
  npm ERR! Lifecycle script `prisma:generate` failed with error: 
  npm ERR! Error: command failed 
  npm ERR!   in workspace: legacy-importer
```

This was the first time after maybe 30 - 40 CI runs, it still feels like enabling the debug output improved things...

**Comment by darthmaim** [comment]
@Jolg42 Is there anything else I can provide to help you narrow down this issue?

**Comment by janpio** [comment]
The only thing that could help is a reliable reproduction that we can run ourselves that let's us experience the problem ourselves.

**Comment by UltimateZero** [comment]
We managed to pinpoint the issue to a possible race condition when running multiple `prisma generate` commands in parallel.
Setting `DEBUG='prisma*'` revealed that both `schema-1` and `schema-2` try to download and write to the same file.

```
schema-1 prisma:generate: 2023-08-25T12:25:08.238Z prisma:fetch-engine:download  Downloading https://binaries.prisma.sh/all_commits/2804dc98259d2ea960602aca6b8e7fdc03c1758f/debian-openssl-1.1.x/libquery_engine.so.node.gz to /home/runner/work/pratiq-microservices/pratiq-microservices/node_modules/prisma/libquery_engine-debian-openssl-1.1.x.so.node 
schema-2 prisma:generate: 2023-08-25T12:25:08.186Z prisma:fetch-engine:download  Downloading https://binaries.prisma.sh/all_commits/2804dc98259d2ea960602aca6b8e7fdc03c1758f/debian-openssl-1.1.x/libquery_engine.so.node.gz to /home/runner/work/pratiq-microservices/pratiq-microservices/node_modules/prisma/libquery_engine-debian-openssl-1.1.x.so.node 
```
(Note: this only occurs if the engine binaries do not exist and need to be downloaded)

Changing our script to run `prisma generate` for each schema sequentially instead of in parallel seems to reliably fix the issue.

**Comment by janpio** [comment]
Uh wow, that command indeed does not make sure it can be run multiple times.

Could this also have been the cause for everyone else here in this issue?

**Comment by jakeleventhal** [comment]
Perhaps, but @janpio we use turborepo which implicitly paralellizes actions. would have to do some acrobatics to set up everything sequentially

**Comment by darthmaim** [comment]
I've setup my generate tasks as dependencies in turborepo, so they should build in sequence. I will report back if the issue is gone over the next few days...

**Comment by darthmaim** [comment]
I did not run into this issue over the last 2 weeks since making sure that `prisma generate` is not run in parallel anymore.

**Comment by jakeleventhal** [comment]
I upgraded from 5.1.1 to 5.3.0 and noticed this error occurring like 90% of the time when running in parallel

**Comment by janpio** [comment]
There were no changes in any of the related code between 5.1.1 to 5.3.0 that we are aware of.

**Comment by jakeleventhal** [comment]
@janpio the problem is really in CI so retrying isn't really an option

**Comment by janpio** [comment]
I know, that was meant about how to reproduce the error - try again until it happens in the reproduction repository. It does not happen that often to me.

**Comment by jakeleventhal** [comment]
FWIW it basically never happens locally but happens much more frequently in CI on default GH actions runners

**Comment by janpio** [comment]
Got the other error case as well now (by running `while clear && find . -type d -name "node_modules" -exec rm -rf {} + && pnpm install; do :; done` in my reproduction repository until I got a different error):

```
│ packages/1 copy 9 prisma-generate: Environment variables loaded from .env
│ packages/1 copy 9 prisma-generate: Prisma schema loaded from prisma/schema.prisma
│ packages/1 copy 2 prisma-generate: Error: ENOENT: no such file or directory, stat '/workspace/prisma-19124/node_modules/.pnpm/prisma@5.3.0/node_modules/prisma/libquery_engine-debian-openssl-3.0.x.so.node'
│ packages/1 copy 2 prisma-generate: Failed
│ undefined
│ /workspace/prisma-19124/packages/1 copy 2:
│  ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  1@1.0.0 prisma-generate: `prisma generate`
│ Exit status 1
└─ Failed in 19.3s at /workspace/prisma-19124
 ELIFECYCLE  Command failed with exit code 1.
```
 
 I think that covers all errors that have been mentioned in this issue.

**Comment by jakeleventhal** [comment]
now the ticket fairy just needs to put a PR under my pillow :)

**Comment by lukahartwig** [comment]
I ran into this issue as well and I this the issue is in [this](https://github.com/prisma/prisma/blob/91ddee5105ffbb81f2336702ca3982e06903b87b/packages/fetch-engine/src/utils.ts#L84) line.

We are running multiple `prisma generate` processes in a monorepo using pnpm workspaces and turborepo. Setting the concurrency to 1 in turborepo resolves the issues.

It looks like the prisma cli deletes and replaces files but not in a concurrency safe way. So other processes that expect those files to be there fail on random fs calls.

**Comment by janpio** [comment]
Multiple Prisma CLI instances indeed might try to access these files at the same time.

Per the comment the workaround via the file deletion, which you identified @lukahartwig, is only needed on MacOS. Would disabling this code on other platforms potentially already help you? That would probably be simple to try out and just add.

**Comment by lukahartwig** [comment]
The problem happens on Ubuntu on GitHub Actions for us. We don't usually run Prisma in parallel locally so making an exception for non-Mac platforms would improve our situation.

**Comment by lukahartwig** [comment]
I just tested wrapping the line in an `os.platform() === "darwin"` check and ran our CI a couple of times. This seems to work. Should I create a PR?

**Comment by Jolg42** [comment]
@lukahartwig feel free to create one, I'll be happy to review it :)

**Comment by jakeleventhal** [comment]
winrar

**Comment by Jolg42** [comment]
@lukahartwig & others: You can now try out this `5.5.0-dev.45` `dev` version. Like `npm install --dave-dev prisma@5.5.0-dev.45 && npm install @prisma/client@5.5.0-dev.45`?

Let us know if that makes things better 🙌🏼 

Note that the official release, 5.5.0 is planned for October 24th.

**Comment by lukahartwig** [comment]
Thanks, I ran our CI three times with the new release and it stays green!

**Comment by Jolg42** [comment]
Then I'll close this 🎊 
Thanks for trying it out!

**Comment by janpio** [comment]
Note: This is potentially leading to problems on Ubuntu that we are currently investigating. We might need to revert. @Jolg42 is on this.

**Comment by Jolg42** [comment]
Note: unfortunately, I had to revert [this](https://github.com/prisma/prisma/pull/21439) temporarily as we spotted a problem in one of our pipelines.

I started to work on https://github.com/prisma/prisma/pull/21480 to actually use the checksum files instead of calling the engine.
Here is a version you can try out to verify that this works well for you:
```
DEBUG=prisma* npm i prisma@5.5.0-integration-joel-debug-download.1
```
Make sure to set `DEBUG=prisma*` to have more debug logs when trying it out.

**Comment by lukahartwig** [comment]
Thanks again. I tested the new version with our CI and it still works.

**Comment by jakeleventhal** [comment]
This issue is really becoming painful for our team. Ruins most CI runs

**Comment by adamspotlite** [comment]
heyo, any update on this? running a turborepo with two different prisma instances and just hit this on our CI. Doesn't seem that there was any consensus on the best path forward?

**Comment by brian-petersen** [comment]
This is becoming increasingly problematic for me too.

**Comment by tonypapousek** [comment]
@Jolg42 / @janpio - I'm not too familiar with Prisma's internal architecture, but could you side-step this issue by moving the queryengine binary download to scripts/install or scripts/postinstall?

**Comment by SadmiB** [comment]
For anyone facing this issue in gitlab ci, the following two lines could help as workaround
```
ENGINE_PATH=$(pnpm prisma version --json | grep -o "node_modules.*libquery_engine-[^)]*")
export PRISMA_QUERY_ENGINE_LIBRARY=$CI_PROJECT_DIR/$ENGINE_PATH
```

**Comment by tmm1** [comment]
@Jolg42 would you mind cutting a dev build off of my PR #25129 that has a potential fix, to make it easier for others here to test?

**Comment by tmm1** [comment]
Friendly ping

**Comment by aqrln** [comment]
@tmm1 

> would you mind cutting a dev build off of my PR #25129 that has a potential fix, to make it easier for others here to test?

Should be available now as `5.21.0-integration-25129.1`. Please install as an exact dependency (i.e. without `^` in `package.json`), otherwise npm may install random stuff that ends up under `integration` tag later.

**Comment by i2equalsj2** [comment]
@aqrln Can confirm that version `5.21.0-integration-25129.1` fixes my pipeline! Ran 10 times without failures whereas before it was failing ~50% of the time.

**Comment by aqrln** [comment]
Cool, I merged the PR, it will be released in Prisma 5.22.0.

**Comment by aqrln** [comment]
Thanks for checking!
